src/Ocean/OceanProblems/initial_value_problem.jl

```diff
@@ -0,0 +1,100 @@
+#####
+##### Initial value problem
+#####
+
+struct InitialValueProblem{FT, IC, BC} <: AbstractSimpleBoxProblem
+    Lˣ::FT
+    Lʸ::FT
+    H::FT
+    initial_conditions::IC
+    boundary_conditions::BC
+
+    """
+        InitialValueProblem{FT}(; dimensions, initial_conditions=InitialConditions(),
+                                boundary_conditions = (OceanBC(Impenetrable(FreeSlip()), Insulating()),
+                                                       OceanBC(Penetrable(FreeSlip()), Insulating())))
+
+    Returns an `InitialValueProblem` with `dimensions = (Lˣ, Lʸ, H)`, `initial_conditions`,
+    and `boundary_conditions`.
+
+    The default `initial_conditions` are resting with no temperature perturbation;
+    the default list of `boundary_conditions` provide implementations for an impenetrable, insulating
+    boundary, and a penetrable, insulating boundary.
+    """
+    function InitialValueProblem{FT}(;
+        dimensions,
+        initial_conditions = InitialConditions(),
+        boundary_conditions = (
+            OceanBC(Impenetrable(FreeSlip()), Insulating()),
+            OceanBC(Penetrable(FreeSlip()), Insulating()),
+        ),
+    ) where {FT}
+
+        return new{FT, typeof(initial_conditions), typeof(boundary_conditions)}(
+            FT.(dimensions)...,
+            initial_conditions,
+            boundary_conditions,
+        )
+    end
+
+end
+
+
+#####
+##### Initial conditions
+#####
+
+resting(x, y, z) = 0
+
+struct InitialConditions{U, V, T, E}
+    u::U
+    v::V
+    θ::T
+    η::E
+end
+
+
+"""
+    InitialConditions(; u=resting, v=resting, θ=resting, η=resting)
+
+Stores initial conditions for each prognostic variable provided as functions of `x, y, z`.
+
+Example
+=======
+
+# A Gaussian surface perturbation
+a = 0.1 # m, amplitude
+L = 1e5 # m, horizontal scale of the perturbation
+
+ηᵢ(x, y, z) = a * exp(-(x^2 + y^2) / 2L^2)
+
+ics = InitialConditions(η=ηᵢ)
+"""
+InitialConditions(; u = resting, v = resting, θ = resting, η = resting) =
+    InitialConditions(u, v, θ, η)
+
+"""
+    ocean_init_state!(::HydrostaticBoussinesqModel, ic::InitialCondition, state, aux, coords, time)
+
+Initialize the state variables `u = (u, v)` (a vector), `θ`, and `η`. Mutates `state`.
+
+This function is called by `init_state_prognostic!(::HydrostaticBoussinesqModel, ...)`.
+"""
+function ocean_init_state!(
+    ::HydrostaticBoussinesqModel,
+    ivp::InitialValueProblem,
+    state,
+    aux,
+    coords,
+    time,
+)
+
+    ics = ivp.initial_conditions
+    x, y, z = coords
+
+    state.u = @SVector [ics.u(x, y, z), ics.v(x, y, z)]
+    state.θ = ics.θ(x, y, z)
```

Comment:
@glwagner would it be better to have

```
state.θ = ics.θ(x, y, z, ivp)
```

instead of 

```
state.θ = ics.θ(x, y, z)
```

and for other ```ics.``` funcs?

For example - what if I want to do an initial condition that is

```
θᵢ(x,y,z,p) = ( 20. * ( 1 - z/p.H ) )
```

Reply:
What's the advantage of being able to reference a parameter as a local variable versus a global variable in the initial conditions function?

If we want functionality like this, I think it's better to attach parameters to `ics.θ` rather than baking this assumption into the function signature. This way we can provide a simple easy-to-use interface, but _support_ more complicated interaction, if desired.

This can be achieved with callable objects:

```julia
struct ParameterizedInitialCondition{P, C}
    condition :: C
    parameters :: P
end

(ivp::ParameterizedInitialCondition)(x, y, z) = condition(x, y, z, parameters)
(ivp::ParameterizedInitialCondition{<:Nothing})(x, y, z) = condition(x, y, z)
```

This way we don't have to change the source code of the function you're referencing; instead we introduce an object that carries around these optional parameters.

On the user-side, this is easy to achieve through dispatch because we already have an interface for `InitialConditions`. So we write add a kwarg `parameters` to `InitialConditions`:

```julia
function InitialConditions(; parameters=nothing, u=default, ...)
    u = ParameterizedInitialCondition(u, parameters)
    # etc
end
```

This way the default is a function signature `x, y, z`, but if you need to pass in parameters, you can.

This pattern is extremely useful for forcing functions and boundary conditions and used extensively in `Oceananigans`. However, generally speaking we haven't found the need to use it for initial conditions --- mainly because initial conditions don't have to run on the GPU and thus don't have special compilation requirements.

Reply:
That works. I don't really need it right now, but we will need something like that - so was just checking.